Lib/test/test_cmd.py

```diff
@@ -289,6 +289,31 @@ def do_tab_completion_test(self, args):
         self.assertIn(b'ab_completion_test', output)
         self.assertIn(b'tab completion success', output)
 
+    def test_bang_completion_without_do_shell(self):
+        script = textwrap.dedent("""
+            import cmd
+            class simplecmd(cmd.Cmd):
+                def completedefault(self, text, line, begidx, endidx):
+                    return ["hello"]
+
+                def default(self, line):
+                    if line == "! hello":
+                        print('tab completion success')
+                    else:
+                        print('tab completion failure')
+                    return True
+
+            simplecmd().cmdloop()
+        """)
+
+        # '! h' and complete 'ello' to '! hello'
+        input = b"! h\t\n"
+
+        output = run_pty(script, input)
+
+        self.assertIn(b'ello', output)
```

Comment:
`ello` because that's what is completed? `hello` is not part of the output?

Reply:
Oh, because that's what is used in the previous test.

Reply:
`ello` because that's what's completed, yes. `hello` is part of the output, but I'm just matching the test above that I copy-pasted from, which only checks for `ab_completion_test`, heh

The tests do both pass if I adjust them to check for `hello` and `tab_completion_test` successfully.

Reply:
I just realized I co-authored that test. Ha.